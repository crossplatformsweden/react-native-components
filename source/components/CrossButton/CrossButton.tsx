import React from 'react';
// @ts-ignore
import { Button } from 'react-native-paper';
import { ViewStyle, StyleProp, View, ButtonProps } from 'react-native';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import { Color } from 'csstype';
import _ from 'lodash';
import styles, { Colors } from '../../styles';

/**
 * Cross button options
 * @typedef ICrossButtonProps
 */
export interface ICrossButtonProps {
  /**
   * Foreground (icon) color. Default is {@link Colors.CrossBlack}
   */
  color?: Color;
  /**
   * Font-Awesome icon name
   *
   * https://fontawesome.com/icons/
   */
  iconName?: string;
  /**
   * Background color. Default value is {@link Colors.NextButton} if {@link title} was supplied
   */
  backgroundColor?: Color;
  style?: StyleProp<ViewStyle>;
  title?: string;
  /**
   * Size of the icon
   */
  size?: number;
  disabled?: boolean;
  /**
   * Button visual appearence. Default is 'text'
   *
   * https://callstack.github.io/react-native-paper/button.html
   */
  mode?: 'text' | 'outlined' | 'contained';
  /**
   * Optional button styles to apply
   */
  buttonStyle?: ButtonProps;
  onPress?: () => void;
  compact?: boolean;
}

/**
 * A custom button that displays as an icon if {@param title} is not supplied
 *
 * Remarks: use custom {@param onPressButton} event
 * @param param0 {@link ICrossButtonProps}
 */
export class CrossButton extends React.Component<ICrossButtonProps> {
  render() {
    const mode = this.props.mode || 'text';
    return (
      <View style={styles.container}>
        {_.isNil(this.props.title) && !_.isNil(this.props.iconName) ? (
          <FontAwesome.Button
            {...this.props}
            style={(this.props.style as ViewStyle) || styles.container}
            onPress={() => {
              if (this.props.onPress) {
                this.props.onPress();
              }
            }}
            color={this.props.color || Colors.CrossBlack}
            name={this.props.iconName.toString()
            }
          />
        ) : (
          <Button
            color={this.props.backgroundColor || Colors.NextButton}
            disabled={this.props.disabled || false}
            {...this.props}
            icon={({ size, color }) =>
              this.props.iconName ? (
                <FontAwesome
                  name={this.props.iconName}
                  size={this.props.size || size}
                  style={{ margin: 0, padding: 0 }}
                  color={color}
                />
              ) : null
            }
            style={this.props.style || this.props.buttonStyle || styles.button}
            onPress={
              this.props.onPress
                ? this.props.onPress
                : () => console.log('** CrossButton press **')
            }
            mode={mode}
          >
            {this.props.title}
          </Button>
        )}
      </View>
    );
  }
}

export default CrossButton;
