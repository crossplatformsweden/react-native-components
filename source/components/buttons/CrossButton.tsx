import React from 'react';
// @ts-ignore
import { Button } from 'react-native-paper';
import { ViewStyle, StyleProp, View, TextStyle } from 'react-native';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import { Color } from 'csstype';
import _ from 'lodash';
import styles, { Colors } from '../../styles';

/**
 * Properties for the {@link CrossButton} component
 * @typedef ICrossButtonProps
 * @type ICrossButtonProps
 */
export interface ICrossButtonProps {
  /**
   * Foreground (icon) color. Default is {@link Colors.CrossBlack}
   */
  color?: Color;
  /**
   * Font-Awesome 4 icon name using `react-native-vector-icons/FontAwesome`
   *
   * https://fontawesome.com/v4.7.0/icons/
   */
  iconName?: string;
  /**
   * Background color. Default value is {@link Colors.NextButton} if {@link title} was supplied
   */
  backgroundColor?: Color;
  /**
   * Style for outer container.
   *
   * See also {@link iconStyle}, {@link buttonStyle}
   */
  style?: StyleProp<ViewStyle>;
  /**
   * Optional style for the icons. See also {@link iconName}
   *
   * See also {@link buttonStyle}, {@link style}
   */
  iconStyle?: ViewStyle | TextStyle;
  /**
   * Title for the button. If not supplied a clickable icon will be displayed instead of react-native-paper button
   *
   * https://callstack.github.io/react-native-paper/button.html
   */
  title?: string;
  /**
   * Size of the icon
   */
  size?: number;
  disabled?: boolean;
  /**
   * Button visual appearence. Default is 'text'
   *
   * https://callstack.github.io/react-native-paper/button.html
   */
  mode?: 'text' | 'outlined' | 'contained';
  /**
   * Optional button styles to apply to {@link https://callstack.github.io/react-native-paper/button.html Button}.
   *
   * See also {@link iconStyle}, {@link style}
   */
  buttonStyle?: StyleProp<ViewStyle>;
  onPress?: () => void;
  compact?: boolean;
}

/**
 * A custom button that displays as an Paper Button with icon or just an icon if {@link ICrossButtonProps.title} is not supplied.
 *
 * Remarks: all icons are {@link https://fontawesome.com/v4.7.0/icons/ FontAwesome v4}.
 *
 * Remark: default appearence is "text", see {@link ICrossButtonProps.mode}
 *
 * Properties are {@link ICrossButtonProps}
 */
export class CrossButton extends React.Component<ICrossButtonProps> {
  render() {
    const mode = this.props.mode || 'text';
    return (
      <View style={this.props.style ? this.props.style : styles.container}>
        {_.isNil(this.props.title) && !_.isNil(this.props.iconName) ? (
          <FontAwesome.Button
            {...this.props}
            style={this.props.iconStyle || styles.container as ViewStyle}
            onPress={() => {
              if (this.props.onPress) {
                this.props.onPress();
              }
            }}
            color={this.props.color || Colors.CrossBlack}
            name={this.props.iconName.toString()}
          />
        ) : (
          <Button
            color={this.props.backgroundColor || Colors.NextButton}
            disabled={this.props.disabled || false}
            {...this.props}
            icon={({ size, color }) =>
              this.props.iconName ? (
                <FontAwesome
                  name={this.props.iconName}
                  size={this.props.size || size}
                  style={{ margin: 0, padding: 0 }}
                  color={color}
                />
              ) : null
            }
            style={this.props.buttonStyle || styles.button}
            onPress={
              this.props.onPress
                ? this.props.onPress
                : () => console.log('** CrossButton press **')
            }
            mode={mode}
          >
            {this.props.title}
          </Button>
        )}
      </View>
    );
  }
}

export default CrossButton;
