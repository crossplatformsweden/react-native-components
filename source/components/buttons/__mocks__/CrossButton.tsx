export const CrossButton = 'Button';

export default CrossButton;
