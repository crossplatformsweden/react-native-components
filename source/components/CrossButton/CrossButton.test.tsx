import React from 'react';
import CrossButton from './CrossButton';
import TestRenderer from 'react-test-renderer';

function setup(iconName: string, title?: any) {
  return TestRenderer.create(
    <CrossButton iconName={iconName} title={title} onPress={jest.fn()} />
  );
}

jest.unmock('react-native');
jest.unmock('./CrossButton');

describe('components', () => {
  /**
   * Test component rendering. Properties of children might be tested by importing their type:
   *
   *    const busyProps = wrapper.find(BusyIndicator).props();
   *    expect(busyProps.isBusy).toBe(false);
   */
  describe('<CrossButton />', () => {
    it('With title should render', () => {
      const wrapper = setup('home', 'home');
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('Without title should renders', () => {
      const wrapper = setup('map', undefined);
      expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });
});
