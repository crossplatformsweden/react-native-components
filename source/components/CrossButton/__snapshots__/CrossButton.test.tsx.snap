// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <CrossButton /> With title should render 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    color="#349b34"
    disabled={false}
    icon={[Function]}
    iconName="home"
    mode="text"
    onPress={[MockFunction]}
    style={
      Object {
        "height": 45,
        "marginRight": 10,
      }
    }
    title="home"
  >
    home
  </View>
</View>
`;

exports[`components <CrossButton /> Without title should renders 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <View
    color="#2f2f2f"
    iconName="map"
    name="map"
    onPress={[Function]}
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
      }
    }
  />
</View>
`;
